scratchwork/fsdp_min/fsdp.py

```diff
@@ -21,51 +22,72 @@
 
 
 def get_fake_data_iter(
-    batch_size: int, vocab_size: int, max_seq_len: int, rank: int, device: torch.device
+    batch_size: int,
+    vocab_size: int,
+    max_seq_len: int,
+    rank: int,
+    device: torch.device,
+    is_validation: bool,
 ) -> Generator[tuple[torch.Tensor, torch.Tensor], None, None]:
     """
-    Fake dataloader. Repeatedly yields the same (inputs, targets) tuple of tensors, with different
-    tensors on different ranks.
+    Fake dataloader. Yields a different set of data for each rank, and for train vs validation.
+    This data would usually come from a tokenized dataset.
     """
-    torch.manual_seed(42 + rank)
-    fake_sequence = torch.randint(vocab_size, (batch_size, max_seq_len), device=device)
-    inputs, targets = fake_sequence[..., :-1], fake_sequence[..., 1:]
+    generator = torch.Generator(device=device)
+    generator.manual_seed(42 + rank + 100000 * is_validation)
     while True:
+        fake_sequence = torch.randint(
+            vocab_size, (batch_size, max_seq_len), device=device, generator=generator
+        )
+        inputs, targets = fake_sequence[..., :-1], fake_sequence[..., 1:]
         yield inputs, targets
 
 
-def get_loss(fsdp_model: FSDP, batch: tuple[torch.Tensor, torch.Tensor]) -> torch.Tensor:
+def get_loss(
+    fsdp_model: FSDP, batch: tuple[torch.Tensor, torch.Tensor], use_amp: bool
+) -> torch.Tensor:
     inputs, labels = batch
-    outputs = fsdp_model(inputs)
-    outputs_flat = outputs.reshape(-1, outputs.shape[-1])
-    labels_flat = labels.reshape(-1)
-    loss = F.cross_entropy(outputs_flat, labels_flat)
+    with torch.cuda.amp.autocast(enabled=use_amp):
+        outputs = fsdp_model(inputs)
+        outputs_flat = outputs.reshape(-1, outputs.shape[-1])
+        labels_flat = labels.reshape(-1)
+        loss = F.cross_entropy(outputs_flat, labels_flat)
     return loss
 
 
 def get_reduced_loss_and_report(
     loss_history: list[torch.Tensor],
     steps_completed: int,
     core_context: det.core.Context,
+    validation: bool,
 ) -> Optional[float]:
     """
-    Average the most recent training losses across all processes and report the result. Returns the
-    reduced loss on rank 0 and None on all other ranks.
+    Average the most recent losses across all processes and report the result.  Returns the reduced
+    loss on rank 0 and None on all other ranks.
     """
 
     loss_history_t = torch.stack(loss_history).mean()
     dist.reduce(loss_history_t, 0, op=dist.ReduceOp.AVG)
     if core_context.distributed.rank == 0:
         reduced_loss = loss_history_t.item()
-        core_context.train.report_training_metrics(
-            steps_completed=steps_completed, metrics={"loss": reduced_loss}
-        )
+        # TypedDict pattern to satisfy mypy.
```

Comment:
`mypy` is so annoying with dicts. I guess something like `report_args: dict[str, Union[int, dict[str, float]]]` or some similar monstrosity would have worked?

Reply:
I'm not sure of a way to do it without `TypedDict` specifically; using a `Union` as the value type doesn't appear to work.  I think that's because `mypy` wants assurance that e.g. `steps_completed: int`, but that only tells it `steps_completed: Union[int, dict[str, float]]`.  So it rejects it since it could be a `dict[str, float]`.

Reply:
Ah, yeah, which makes sense, but is annoying regardless.